### ISSUE TITLE
Event Manager

### ISSUE THREAD

**Issue by lavalamp** [kubernetes/kubernetes#3329]
Now that we have various components producing events, we need to start processing them. Our current policy of a 2 day TTL isn't going to scale.

So, I want us to build an event manager.

The basic sketch is that it reads events from the cluster, processes them, and archives them.

a. Reading events:
1. It runs in a pod as part of your kubernetes cluster.
2. It uses the "kubernetes" service to contact the master.
3. It lists & then watches all events.

b. Process events
1. Idea: compression. If you see N events identical except for timestamp, transform that into one event with N timestamps.
2. Alternatively, do the same but for sequences of events. Like if a pod is in a pull -> fail -> pull... loop, make a meta-event and start storing timestamps.
3. Main purpose: Actively manage the events stored in the kubernetes system, with the goal of maximizing usefulness while storing only a sane number of events.
4. Idea: provide scriptable/configurable hooks. For example, it would great for an admin to make a policy here that deletes a pod after N failures to start. Or up a level, to delete a replication controller that's making crashlooping pods. Or delete a pod if it fails to schedule after 30 minutes. Etc.
5. Idea: provide scriptable/configurable alerts. As above, but email/page an admin/owner instead of deleting something.
6. Idea: upon reading an event, update the object's status information to reflect the new information.

c. Archive events
1. Store the events in a DB of some sort for offline analysis. Ideally the latency between the event getting added to k8s and getting stored in the DB is low

d. Misc open questions
1. Run one per namespace or one per cluster? I'd prefer the latter, but we'd have to be cognizant of not leaking information between namespaces.

We can also explore the idea of kubernetes writing events to a non-etcd (or separate etcd) DB in the first place. Events are the primary source of write load on etcd right now. It's good for the moment in that it's exposing some bugs in our use of etcd, but in the long term it's probably more efficient for us to use a different storage mechanism for events.

**Comment by davidopp** [comment]
How do we decide what information to put into PodStatus/NodeStatus and what information to put into discrete events? I believe there is some kind of homomorphism here because in the limit we could
- embed events in the PodStatus/NodeStatus and remove it as a discrete API type; or on the other extreme
- remove PodStatus/NodeStatus as API types and make them virtual objects that the client constructs by observing events

Also, on the archiving question -- could/should we just archive the whole etcd transaction log? (I assume etcd is structured as some kind of log that contains every mutating operation, perhaps with some kind of periodic compaction) This could be useful for post-hoc debugging and would give us archiving of events for free (if you were only interested in events when reading back, you'd skip all the non-event mutations).

**Comment by lavalamp** [comment]
> How do we decide what information to put into PodStatus/NodeStatus and what information to put into discrete events?

XStatus should have the entire current state of X; an event about X tells you only one detail. I think there is room for both. I added b.6. above to capture the idea of updating status based on incoming events. It's not clear if that's actually the best thing to do, but it's worth talking about.

> Also, on the archiving question -- could/should we just archive the whole etcd transaction log?

This is a fair point-- perhaps we should solve archival globally. But it would be good to store it in a searchable/queryable format.

**Comment by derekwaynecarr** [comment]
I prefer that we solve archival globally, and if we use events as the first use case, that is fine for me.

----- Original Message -----

**Comment by dchen1107** [comment]
#2298 too

We plan to moving XStatus (at least PodStatus) computation to Kubelet level, which means keeping XStatus as API type is necessary.

**Comment by davidopp** [comment]
One possible model for Status vs. Events is the following (based on a similar system I worked on recently). I think this is similar to what you were saying.

A component that starts with no information can construct the entire state of the cluster by observing the current values of the Status objects in etcd. If it later becomes backlogged or goes down for a long time, it can come back up and reconstruct the state this way. On the other hand, Event lifetimes are bounded (either by an explicit TTL, or the compaction interval of the transaction log, or the retention policy of the system that archives the transaction log, or whatever). So Events should only be used to convey information that is not mission-critical (for example, you specifically would not want to delegate conversion of Events into Status to an event manager that runs on top of Kubernetes, since losing some Events would corrupt the representation of the cluster's state; having Kubelet and API server compute Status is safest).

This also provides a design guideline for components, which in the previous project we described as "edged-based" vs. "level-based"; the former rely on seeing every object transition, while the latter could figure out what to do just based on its observation of the current Statuses and whatever private state it had stashed away. Unfortunately by the time we understood this distinction we were already building edge-based components so we just pretended the store transaction log would be archived long enough that even if the edge-based component was down (or fell behind) for a long time enough time it would always be able to catch up. This isn't a good/safe assumption to make.

**Comment by dchen1107** [comment]
We extracted work items required for v1 release, and filed them separately. Lower the priority of this to P3 to unblock V1.

**Comment by smarterclayton** [comment]
This is highly theoretical and in practice isn't an urgent issue.  Please re-open if you disagree.

**Comment by falenn** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  curtis.l.bates@gmail.com
    Domain imwiz.com has exceeded the max emails per hour (158/150 (105%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from o9.sgmail.github.com ([167.89.101.2]:20471)
	by box969.bluehost.com with esmtps (TLSv1.2:ECDHE-RSA-AES128-GCM-SHA256:128)
	(Exim 4.87)
	(envelope-from <bounces+848413-5c7e-dev=imwiz.com@sgmail.github.com>)
	id 1dGTL9-000Tsb-Jb
	for dev@imwiz.com; Thu, 01 Jun 2017 10:52:52 -0600
DKIM-Signature: v=1; a=rsa-sha1; c=relaxed/relaxed; d=github.com; 
	h=from:reply-to:to:cc:in-reply-to:references:subject:mime-version:content-type:content-transfer-encoding:list-id:list-archive:list-post:list-unsubscribe; 
	s=s20150108; bh=sWjMG0zxujO+bliemWP/guo18C8=; b=fZfNG3rHkyZkI1jM
	bCn2xBtv+PcgvpRtglDDoJPDY1TL68jca5Rd0GmELtqMbNtsD4tjbbmigXc4BWPl
	usDVJd2Fa+lG/QqPn8cV3OIv5fpOiHajo9VWPKdRTgV1B1Ud3XuKKLn9sCIovPY6
	u90hpuymsiIx7FVnCtPaU5U/I0U=
Received: by filter0501p1mdw1.sendgrid.net with SMTP id filter0501p1mdw1-2636-59304654-6A
        2017-06-01 16:52:36.983492999 +0000 UTC
Received: from github-smtp2a-ext-cp1-prd.iad.github.net (github-smtp2a-ext-cp1-prd.iad.github.net [192.30.253.16])
	by ismtpd0006p1iad1.sendgrid.net (SG) with ESMTP id yNz9ELt9QHK39MwvLPFGGw
	for <dev@imwiz.com>; Thu, 01 Jun 2017 16:52:36.952 +0000 (UTC)
Date: Thu, 01 Jun 2017 09:52:36 -0700
